File: CSharp/Location.cs
namespace ThingModel
{
    public abstract class Location
    {
        public double X;
        public double Y;
        
        /** This parameter is optionnal **/
        public double? Z;

        /**
         * Name of the location system.
         * 
         * Examples, WGS84, ETRS89...
         */
        public string System;
        
        /**
         * You cannot use this constructor directly.
         *  Location.{Point|LatLng|Equatorial} should be used instead.
         */
        protected Location(double x, double y, double? z)
        {
            X = x;
            Y = y;
            Z = z;
        }

        /**
         * Return true if the location contain the same values.
         * 
         * A Location.Point and a Location.LatLng are the same
		 * if the values are the same.
         * 
         * The comparaison doesn't manage issues with precision numbers. If
         * a value is rounded, the location is not the same.
         */
        public bool Compare(Location other)
        {
            // Double comparaison without Epsilon mouahahaha
            // ReSharper disable CompareOfFloatsByEqualityOperator
            return other != null &&
				X == other.X &&
				Y == other.Y &&
				Z == other.Z &&
				System == other.System;
            // ReSharper restore CompareOfFloatsByEqualityOperator
        }

        /**
         * Used for videogames and for default coordinates systems.
         */
        public class Point : Location
        {
            public Point(double x = 0.0, double y = 0.0, double? z = null)
                : base(x, y, z)
            {
            }
        }

        /**
         * For location on planets.
         * 
         * Will be used often with Earth WGS-84 (GPS).
         */
        public class LatLng : Location
        {
            public LatLng(double latitude = 0.0, double longitude = 0.0, double? altitude = null)
                : base(latitude, longitude, altitude)
            {
            }

            public double Latitude
            {
                get { return X; }
                set { X = value; }
            }

            public double Longitude
            {
                get { return Y; }
                set { Y = value; }
            }

            public double? Altitude
            {
                get { return Z; }
                set { Z = value; }
            }
        }

        /**
         * For location in space.
         */
        public class Equatorial : Location
        {
            public Equatorial(double rightAscension = 0.0, double declination = 0.0, double hourAngle = 0.0)
                : base(rightAscension, declination, hourAngle)
            {
            }

            public double RightAscension
            {
                get { return Y; }
                set { Y = value; }
            }

            public double Declination
            {
                get { return X; }
                set { X = value; }
            }

            public double HourAngle
            {
                get { return Z != null ? (double) Z : 0.0; }
                set { Z = value; }
            }
        }


	    protected bool Equals(Location other)
	    {
		    return X.Equals(other.X) && Y.Equals(other.Y) && Z.Equals(other.Z) && string.Equals(System, other.System);
	    }

	    public override bool Equals(object obj)
	    {
		    if (ReferenceEquals(null, obj)) return false;
		    if (ReferenceEquals(this, obj)) return true;
		    if (obj.GetType() != GetType()) return false;
		    return Equals((Location) obj);
	    }

	    public override int GetHashCode()
	    {
		    unchecked
		    {
// ReSharper disable NonReadonlyFieldInGetHashCode
			    int hashCode = X.GetHashCode();
			    hashCode = (hashCode*397) ^ Y.GetHashCode();
			    hashCode = (hashCode*397) ^ Z.GetHashCode();
			    hashCode = (hashCode*397) ^ (System != null ? System.GetHashCode() : 0);
// ReSharper restore NonReadonlyFieldInGetHashCode
			    return hashCode;
		    }
	    }
    }
}

File: TypeScript/Location.ts
module ThingModel {
	export interface Location {
		X: number;
		Y: number;
		Z: number;

		/**
		 *	Name of the location system.
		 *
		 *	Examples : WGS84, ETRS89...
		 */
		System: string;

		/**
		 *	Return true if the locations are the same.
		 */
		Compare(other: Location): boolean;

		toString(): string;

		type: string;
	}

	export module Location {
		/**
		 *	Simple location value.
		 *	Often used in videogames.
		 */
		export class Point implements Location {
			public X: number;
			public Y: number;
			public Z: number;
			public System: string;
			public type: string;

			// The last argument can be null, it's a feature
			constructor(x: number = 0.0, y: number = 0.0, z: number = null) {
				this.X = x;
				this.Y = y;
				this.Z = z;
				this.type = "point";
			}

			// Check if the values are the same and if the representation is the same
			public Compare(other: Location) {
				return other != null &&
					other.X === this.X &&
					other.Y === this.Y &&
					other.Z == this.Z &&
					other.System == this.System;
			}

			public toString() {
				var s = this.X + " - " + this.Y;

				if (this.Z != null) {
					s += " - " + this.Z;
				}

				if (this.System != null) {
					s += " -- " + this.System;
				}
				return s;
			}
		}

		/**
		 *	Latitude longitude representation, often used for WGS 84 GPS localizations.
		 *	And if the System property is null, it's considered by default as a WGS 84 LatLng system.
		 */
		export class LatLng extends Point {

			constructor(latitude: number = 0.0, longitude: number = 0.0, altitude: number = null) {
				super(latitude, longitude, altitude);
				this.type = "latlng";
			}

			public get Latitude() {
				return this.X;
			}

			public set Latitude(latitude: number) {
				this.X = latitude;
			}

			public get Longitude() {
				return this.Y;
			}

			public set Longitude(longitude: number) {
				this.Y = longitude;
			}

			public get Altitude() {
				return this.Z;
			}

			public set Altitude(altitude: number) {
				this.Z = altitude;
			}

		}

		/**
		 *	Representations in space.
		 */
		export class Equatorial extends Point {

			constructor(rightAscension: number = 0.0, declination: number = 0.0, hourAngle: number = 0.0) {
				super(rightAscension, declination, hourAngle);
				this.type = "equatorial";
			}

			public get RightAscension() {
				return this.X;
			}

			public set RightAscension(rightAscension: number) {
				this.X = rightAscension;
			}

			public get Declination() {
				return this.Y;
			}

			public set Declination(declination: number) {
				this.Y = declination;
			}

			public get HourAngle() {
				return this.Y;
			}

			public set HourAngle(hourAngle: number) {
				this.Z = hourAngle;
			}

		}
	}
}